### PR TITLE
Store comments in datastore

### PR DESCRIPTION
**Pull request by mes712**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -22,30 +22,51 @@
...
 import java.util.List;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.common.collect.ImmutableList;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.common.collect.Streams;
+import static com.google.common.collect.ImmutableList.toImmutableList;
```
Comment (linhnguye-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: static imports should be ordered before non-static imports

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -22,30 +22,51 @@
...
-     Gson gson = new Gson();
-     String json = gson.toJson(comments);
-     response.getWriter().println(json);
-   }
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json;");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    ImmutableList<Comment> comments = Streams.stream(results.asIterable())
```
Comment (linhnguye-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: super nit: more readable to move this to next line:

```
ImmutableList<Comment> comments =
    Streams.stream(results.asIterable())
        .map(...)
        .collect(...);
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -22,30 +22,51 @@
...
 
    @Override
-   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-     response.setContentType("application/json;");
-     Gson gson = new Gson();
-     String json = gson.toJson(comments);
-     response.getWriter().println(json);
-   }
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json;");
```
Comment (linhnguye-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: move this down to group together with `response.getWriter().println(json);`

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -34,7 +34,9 @@ function changeCatPhoto() {
 /* Displays all comments left on page. */
 async function getComments() {
   const response = await fetch('/data').then(response => response.json())
+  console.log(response);
```
Comment (linhnguye-g) [inline portfolio/src/main/webapp/script.js]: nit: remove logs, same with below

**Review by linhnguye-g — APPROVED** [review]
Nice use of streams :)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -22,30 +22,51 @@
...
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String comment = getParameter(request, COMMENT_NAME, DEFAULT_VAL);
+    String commentText = getParameter(request, COMMENT_NAME, DEFAULT_VAL);
     String displayName = getParameter(request, DISPLAY_NAME, DEFAULT_VAL);
+    long timestamp = System.currentTimeMillis();
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("comment", commentText);
+    commentEntity.setProperty("name", displayName);
+    commentEntity.setProperty("timestamp", timestamp);
```
Comment (christophervo-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: same thing about using constants for the property names. We can reuse some of the ones we defined for the earlier comment.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -22,30 +23,51 @@
...
-     String json = gson.toJson(comments);
-     response.getWriter().println(json);
-   }
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    ImmutableList<Comment> comments = 
+        Streams.stream(results.asIterable())
+            .map(entity -> (new Comment((String) entity.getProperty("comment"), (String) entity.getProperty("name"))))
```
Comment (christophervo-g) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: Nice job on the streams! For readability, let's separate this into a helper method that takes an Entity and returns a new Comment. Also, let's define "comment" and "name" as constants (maybe something like COMMENT_PROPERTY and NAME_PROPERTY?)

**Review by christophervo-g — APPROVED** [review]
LGTM, just some minor nits

**Pull request merged**